### PR TITLE
Re-Add Auto-Safe-Mode Coloring

### LINKED ISSUE #19467: Safemode indicator no longer changes color
The Safemode indicator will always stay green no matter what in 0.C-19885-g4a966d2 (tiles)/Build 5945 (according to the CDDA Launcher). Previously, the Safemode indicator would turn entirely red when Safemode was disabled, and parts of the safemode indicator would gradually turn green again as the number of turns left before safemode was auto-enabled decreased.

To recreate this, have Auto Safe Mode, Safe Mode, and Turns to Re-Enable Safe Mode (Probably set it to 20 or something) enabled in the options, and then run around with safemode disabled while trying to see if the indicator will change colors at all.

### LINKED ISSUE #19597: Auto-Safe Mode not sticking when enabled/disabled in-game.
Win10 x64 0.C-20040-g20172c1 (#5995)(tiles)

Changing Auto-Safe mode in-in-game (either through settings or keybind) doesn't seem to work, only changes during the main menu/pre-load seem to stick. ie With Auto-Safe mode disabled in settings before starting a game, it will always be disabled, even if enabled in-game.

Related #19467 (to my original title)

### PR DESCRIPTION
**Pull request by tswood**
fixes #19467 
fixes #19597

**Main Goals**
* Re-implement per-character coloring of safe mode indicato for autosafemode duration progression.
* Fix toggling autosafemode via options while in-game.

**Implementation Details**
* Extracted safe mode rendering to new `draw_safe_mode( WINDOW *win, int line ) const` method.
* Wrapped translated safe text in `utf8_wrapper` to allow for easier char manipulation.
* Removed `game::autosafemode` as it was only updated via load and in-game keybound toggle, replaced with getting the latest option available.

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -5279,6 +5277,36 @@ void game::draw_sidebar()
...
+    }
+
+    const char *safe_text = _( "SAFE" );
+    if (safe_mode != SAFE_MODE_OFF) {
+        right_print( w_location, 0, 1, c_green, "%s", safe_text );
+        return;
+    }
+
+    if (autosafemode) {
+        int text_length = strlen( safe_text );
```
Comment (illi-kun) [inline src/game.cpp]: You need to use ```utf8_width()``` and support utf8 in general.

**Comment by BrettDong** [comment]
Would be better if format your new code in the rules of `.astylerc`

**Comment by tswood** [comment]
Added support for UTF-8 strings. Compiled with all languages and tested all available languages on OSX.

Apply astyle formatting to new function.

Thanks for the feedback!

File: src/game.cpp [hunk 2]
```diff
@@ -5279,6 +5277,41 @@ void game::draw_sidebar()
     draw_sidebar_messages();
 }
 
+void game::draw_safe_mode()
```
Comment (mugling) [inline src/game.cpp]: Should expect a `WINDOW *, int y, int x` as arguments as opposed to hardcoding them

Comment (tswood) [inline src/game.cpp]: Passing in the window to allow for decoupling there makes sense. As for passing in the x/y, would that not require pulling out some of the internal logic happening in the `draw_safe_mode()` function? The y / line perhaps, but would the x represent the left-aligned start or right-aligned start? If left aligned, I would think then the calling code would already have to know about the translated SAFE string and length, no?

**Comment by tswood** [comment]
Moved window and at least line / y to parameters of the new method as well as added some missing consts. If there is a good way to also allow dynamic positioning (left or right aligned) that is felt necessary for this at this time, I can make that change as well. At this time and prior however the text has been right-aligned with an offset of 1.

File: src/game.cpp [hunk 3]
```diff
@@ -5279,6 +5277,41 @@ void game::draw_sidebar()
...
+        const float safe_mode_percent =
+            turnssincelastmon * 100.00 / get_option<int>( "AUTOSAFEMODETURNS" );
+
+        const int text_size = safe_text.size();
+        const int starting_position = getmaxx( win ) - safe_text.display_width() - 1;
+
+        int written_size = 0;
+        for( int i = 0; i < text_size; i++ ) {
+            nc_color letter_color =
+                safe_mode_percent < ( i + 1 ) * ( 100.00 / text_size )
```
Comment (mugling) [inline src/game.cpp]: Minor: Too much vertical paging. Should have been one or two lines

**Comment by tswood** [comment]
Updated for formatting and retested on linux64 ncurses build. Thanks.

**Comment by mugling** [comment]
See comments in #19597

**Comment by tswood** [comment]
PR has been updated to also address #19597. Updated original post.

**Pull request merged**